File: src/css.cc
/*
 * File: css.cc
 *
 * Copyright 2008-2014 Johannes Hofmann <Johannes.Hofmann@gmx.de>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 */

#include "css.hh"
#include "css.h"

CssLengthType cpp_cssLengthType(CssLength len)
{
   return (CssLengthType) len.length_type;
}

float cpp_cssLengthValue(CssLength l)
{
   return (float) l.length_value;
}

CssLength cpp_cssCreateLength(float v, CssLengthType t)
{
   CssLength retv = { .length_value = (double) v, .length_type = (int) t };
   return retv;
}




/*
 Parsing code, moved from cssparser.cc.

 This is the code that downloads stylesheet specified by @import.

 TODO: use it in the future.
*/
#if 0
void parseImport(DilloHtml *html, c_css_parser_t * parser, c_css_token_t * token, const DilloUrl * base_url)
{
   if (urlStr) {
      if (importSyntaxIsOK && mediaIsSelected) {
         MSG("CssParser::parseImport(): @import %s\n", urlStr);
         DilloUrl *url = a_Html_url_new (html, urlStr, a_Url_str(base_url),
                                         base_url ? 1 : 0);
         a_Html_load_stylesheet(html, url);
         a_Url_free(url);
      }
      dFree (urlStr);
   }
}
#endif

File: src/Hello/Css/Css.hs
{-
Copyright (C) 2021 Kamil Ignacak acerion@wp.pl

This file is part of "hello" web browser.

"hello" is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

"hello" is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with "hello".  If not, see <https://www.gnu.org/licenses/>.

This file is derived from dillo-3.0.5/src/css.cc (and doctree.hh).
Copyright assignments from css.cc file:
Copyright 2008-2014 Johannes Hofmann <Johannes.Hofmann@gmx.de>
-}


module Css
  (
    selectorSpecificity
  )
  where




import qualified Data.Text as T
import qualified Data.Map as M
import Data.Bits
import Debug.Trace
import Hello.Css.Parser
import Hello.Css.Selector
import Hello.Css.DoctreeNode
import Hello.Ffi.Css.DoctreeNode





{-
Return the specificity of the selector.

The specificity of a CSS selector is defined in
http://www.w3.org/TR/CSS21/cascade.html#specificity
-}
selectorSpecificity :: CssComplexSelector -> Int
selectorSpecificity complex = selectorSpecificity' complex 0
  where
    selectorSpecificity' :: CssComplexSelector -> Int -> Int
    selectorSpecificity' (Link (Datum c1) combinator remainder) acc = selectorSpecificity' remainder (acc + (compoundSelectorSpecificity c1))
    selectorSpecificity' (Datum c1)                             acc =                                 acc + (compoundSelectorSpecificity c1)




-- Return the specificity of compound selector
compoundSelectorSpecificity :: CssCompoundSelector -> Int
compoundSelectorSpecificity compound = (fromId compound) + (fromClass compound) + (fromPseudoClass compound) + (fromElement compound)
  where
    fromId compound          = if (not . T.null . selectorId $ compound) then (1 `shiftL` 20) else 0
    fromClass compound       = (length . selectorClass $ compound) `shiftL` 10
    fromPseudoClass compound = if (not . null . selectorPseudoClass $ compound) then (1 `shiftL` 10) else 0 -- Remember that C/C++ code can use only first pseudo code.
    fromElement compound     = if compoundHasUniversalType compound then 0 else 1




